_posts/2021-01-05-overlappable-instances.markdown

```diff
@@ -0,0 +1,248 @@
+  => Store m
+  => m ()
+program = do
+  put key1 val1
+  log $ "Inserted value: " <> show val1 <> " under key: " <> key1
+  maybeV <- get key1
+  case maybeV of
+    Just v -> log $ "Retrieved: " <> v
+    Nothing -> log "No data found"
+  where
+    key1 = "key1"
+    val1 :: Integer
+    val1 = 1
+
+instance Store IO where
+  put _ _ = pure ()
+  get _ = pure Nothing
+
+instance Logger IO where
+  log _ = pure ()
+
+main :: IO ()
+main = program
+```
+This is clearly limiting, because we are allowed to have only one typeclass instance for a given monad. If we would like to have a NoOp instance and a real-world instance, we have to use newtype wrappers. Let's add aliases then for our not-so-useful instances like this:
+
+```haskell
+newtype NoOpStoreT m a = NoOpStoreT { runNoOpStoreT :: m a }
+  deriving newtype (Functor, Applicative, Monad)
+
+instance Monad m => Store (NoOpStoreT m) where
+  put _ _ = pure ()
+  get _ = pure Nothing
+
+newtype NoOpLoggerT m a = NoOpLoggerT { runNoOpLoggerT :: m a }
+  deriving newtype (Functor, Applicative, Monad)
+
+instance Monad m => Logger (NoOpStoreT (NoOpLoggerT m)) where
+  log _ = pure ()
+
+main :: IO ()
+main = runNoOpLoggerT . runNoOpStoreT $ program
+
+```
+This compiles and runs fine, but we've arrived at the infamous O(n^2) instances problem! If elsewhere in our codebase we stack our transformers in a different order, e.g. `runNoOpStoreT . runNoOpLoggerT $ program` we will have to add new instances for the transformers' stack in that particular order. Not nice!
+```
+    • No instance for (Logger (NoOpLoggerT (NoOpStoreT IO)))
+        arising from a use of ‘program’
+```
+Overlappable transformer instances to the rescue! Adding these both for `Store` and `Logger` makes it possible to get rid of this problem and allow us to stack our transformers in any* order! Let's start with `Logger`:
+
+```haskell
+instance MonadTrans NoOpStoreT where
+  lift a = NoOpStoreT a
+
+instance (Logger m, MonadTrans t, Monad m) => Logger (t m) where
+  log a = lift $ log a
+
+instance Monad m => Logger (NoOpLoggerT m) where
+  log _ = pure ()
+```
+We've defined a new instance. It is going to be available only if when a `Logger` instance already exists for `m` and we know how to wrap this monad `m` using `t`. For this to be true we need to add a `MonadTrans` instance for our `NoOpStoreT`. Thanks to this we don't have to write all n^2 instances by hand. But the compiler has a problem now:
+
+```
+   • Overlapping instances for Logger (NoOpLoggerT IO)
+        arising from a use of ‘program’
+      Matching instances:
+        instance (Logger m, MonadTrans t, Monad m) => Logger (t m)
+          -- Defined at ...
+        instance Monad m => Logger (NoOpLoggerT m)
+          -- Defined at ...
+
+```
+We need to tell it, that whenever two instances could be used "arising from the use of program", one of them is less specific, and labeled using the `{-# OVERLAPPABLE #-}` directive. Now the code compiles fine.
```

Comment:
yes!
Minor point: the [GHC User's Guide](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/glasgow_exts.html?highlight=templatehaskellquotes#overlapping-overlappable-overlaps-and-incoherent-pragmas) calls these _pragmas_, not _directives_